pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,244 @@ var (
+			rd.Allows[requirement.Key] = rd.Allow(requirement.Key).Intersection(sets.NewComplementSet(requirement.Values...))
+		}
+	}
+	return rd
 }
 
-func (r Requirements) CapacityTypes() sets.String {
-	return r.Requirement(LabelCapacityType)
+func (r *Requirements) MarshalJSON() ([]byte, error) {
+	var result []v1.NodeSelectorRequirement
+	for _, requirements := range r.Records {
+		result = append(result, requirements...)
+	}
+	return json.Marshal(result)
 }
 
-func (r Requirements) Add(requirements ...v1.NodeSelectorRequirement) Requirements {
-	return append(r, Requirements(requirements).Normalize()...)
+func (r *Requirements) UnmarshalJSON(b []byte) error {
+	var requirements []v1.NodeSelectorRequirement
+	json.Unmarshal(b, &requirements)
+	result := NewRequirements(requirements...)
+	r.Records = result.Records
+	r.Allows = result.Allows
+	return nil
 }
 
-// Normalize the requirements to use WellKnownLabels
-func (r Requirements) Normalize() Requirements {
-	normalized := Requirements{}
-	for _, requirement := range r {
-		label := requirement.Key
-		if normalized, ok := NormalizedLabels[requirement.Key]; ok {
-			label = normalized
-		}
-		normalized = append(normalized, v1.NodeSelectorRequirement{Key: label, Operator: requirement.Operator, Values: requirement.Values})
+// Allow returns the sets of values allowed by all included requirements
+// Allow follows a denylist method. Values are allowed except specified
+func (r *Requirements) Allow(key string) sets.Set {
+	if r == nil {
+		return sets.NewComplementSet()
 	}
-	return normalized
+	if _, ok := r.Allows[key]; !ok {
+		return sets.NewComplementSet()
+		//r.Allows[key] = sets.NewComplementSet()
+	}
+	return r.Allows[key]
 }
 
-// Consolidate combines In and NotIn requirements for each unique key, producing
-// an equivalent minimal representation of the requirements. This is useful as
-// requirements may be appended from a variety of sources and then consolidated.
-// Caution: If a key has contains a `NotIn` operator without a corresponding
-// `In` operator, the requirement will permanently be [] after consolidation. To
-// avoid this, include the broadest `In` requirements before consolidating.
-func (r Requirements) Consolidate() (requirements Requirements) {
+// Validate validates the feasibility of the requirements.
+func (r *Requirements) Validate() (errs *apis.FieldError) {
+	if r == nil {
+		return errs
+	}
 	for _, key := range r.Keys() {
-		requirements = requirements.Add(v1.NodeSelectorRequirement{
-			Key:      key,
-			Operator: v1.NodeSelectorOpIn,
-			Values:   r.Requirement(key).UnsortedList(),
-		})
+		for _, err := range validation.IsQualifiedName(key) {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("%s, %s", key, err), "key"))
+		}
+		values := r.Allow(key)
+		if values.Len() == 0 {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("no feasible value for requirement, %s", key), "values"))
+		}
+		i := 0
+		for value := range values.Members {
```

Comment:
consider using something like `i, value := range values.Members.UnsortedList()`

Reply:
So we should index the errors using the requirement index instead?
Since the requirement index will keep changing for UnsortedList(), doesn't that make the index inconsistent?

Also are we allowed to have a gap between error indexes?